Add some core::cmp::Ordering helpers

jnqnfe opened the pull request:
...to allow easier equal-to-or-greater-than and less-than-or-equal-to
comparisons.

Prior to Rust 1.42 a greater-than-or-equal-to comparison might be written
either as a match block, or a traditional conditional check like this:

```rust
if cmp == Ordering::Equal || cmp == Ordering::Greater {
    // Do something
}
```

Which requires two instances of `cmp`. Don't forget that while `cmp` here
is very short, it could be something much longer in real use cases.

From Rust 1.42 a nicer alternative is possible:

```rust
if matches!(cmp, Ordering::Equal | Ordering::Greater) {
    // Do something
}
```

The commit adds another alternative which may be even better in some cases:

```rust
if cmp.is_equal_or_greater() {
    // Do something
}
```

The earlier examples could be cleaner than they are if the variants of
`Ordering` are imported such that `Equal`, `Greater` and `Less` can be
referred to directly, but not everyone will want to do that.

The new solution can shorten lines, help avoid logic mistakes, and avoids
having to import `Ordering` / `Ordering::*`.

rust-highfive commented:
r? @sfackler

(rust-highfive has picked a reviewer for you, use r? to override)

jnqnfe commented:
If `is_equal()`, `is_greater()` and `is_less()` helpers are wanted also, I'm happy to add incorporate those too.

egilburg commented:
Bikeshedding: Common word syntax for such matchers is `eq`, `neq` `gt`, lt`, `gte`, `lte`. Perhaps add them as such?

sfackler commented:
I agree that `is_lt`, `is_le`, `is_eq`, etc make sense to match the `PartialOrd` method names.

scottmcm commented on library/core/src/cmp.rs:
````diff
@@ -441,6 +441,44 @@ impl Ordering {
+    /// ```
+    #[inline]
+    #[must_use]
+    pub const fn is_equal_or_greater(self) -> bool {
````
nit: given that it's `PartialOrd::ge`, consider `is_greater_or_equal`.

jnqnfe replied:
It's now `is_ge`

scottmcm replied:
Oh, it didn't already have `is_equal` and such for single variants.  `is_ge` sounds much better than my suggestion, then.

jnqnfe commented:
Ok, I've revised things, including addressing the feedback.
 - Added the additional three helpers mentioned along with a forth (not-equal).
 - Adjusted to the shorter names suggested. (`is_eq`, `is_ne`, `is_lt`, `is_gt`, `is_le`, `is_ge`).
 - Moved them to the start of the set of methods for the type, since they are more fundamental than those that already exist.
 - Revised the commit log.

I force-pushed the changes to keep things tidy, since it's such a trivial one to review.

scottmcm commented:
@jnqnfe Looks like this is still failing CI?

jnqnfe commented:
> @jnqnfe Looks like this is still failing CI?

Yeah, the failure is stating that it 'could not download the log' and to try again later... I see no button to ask it to try again...

scottmcm commented:
You have a bunch of these:
![image](https://user-images.githubusercontent.com/18526288/101698568-9cb8bc00-3a71-11eb-9d73-9ff8ad887232.png)

It'll probably have to use `matches!` for all of them.

jnqnfe commented:
> You have a bunch of these:
> [snip]
> It'll probably have to use `matches!` for all of them.

Ah, thanks, fixed! (Sorry about that, it's been a while since I last contributed and I haven't been able to get things setup for compiling this time, partly because I'm using the stable compiler version from my repo :/ )

I've force-pushed that fix, fingers crossed CI will work this time :)

Oh, result came in before I could post, it still can't get the log...

jnqnfe commented:
@scottmcm Ohhhhh, I've just noticed that a github domain was being blocked by noscript in my browser and as a consequence the errors you saw were being hidden from me. I was under the impression that there was some issue with CI itself because all I saw was the 'could not download log' error line. Sorry about that, I would have fixed the errors days ago if I'd realised.

I've just fixed the typo in the last push, so should be good now.

Okay, so now it's complaining about missing stability attributes...

jnqnfe commented:
@scottmcm Looking into what stability attribute to assign, I'm not at all certain.

 - Do I need to go through the effort of creating a new unique feature gate for such a simple addition as this? Or can I use something like `std_misc` as the feature?
 - Can I mark this as stable (for 1.50.0?) immediately, or do you want it marked unstable?
 - If marking unstable, must I then create a tracking issue for this or is it acceptable to set the issue attribute to `"none"`?

scottmcm commented:
See https://rustc-dev-guide.rust-lang.org/stability.html#unstable

Yes, you need a feature name and a tracking issue.  Just invent a reasonable name; you don't need to put it anywhere other than in the attributes.

Whenever possible -- as it is here -- things go in unstable.  That allows them to go in with just a PR sign-off, whereas stabilization will require a full libs team FCP.

jnqnfe commented:
@scottmcm Ok. I've:
 - Created the tracking issue.
 - Created a new feature `ordering_helpers` in `rustc_features` (that guide btw needs updating to note that this exists under `compiler/`).
 - Created an unstable book entry, similar to others.
 - Tagged the new methods with the appropriate attribute.
 - Added a "sym" entry in `rustc_span`, following a CI error and guessing from other merged work. (This was not mentioned at all in the contributing guide).
 - Added a test, making a complete guess at where from copying other merged work.

Outstanding:
 - Filling in the `stderr` file for the test. Since I don't have the nightly compiler it seems I can't compile to get it. Can you please help with this.

I merged the changes into a single commit. I'll split things up if preferred. Thanks for your patience.

scottmcm commented:
Sorry, I guess I wasn't clear here.  The "you don't need to put it anywhere other than in the attributes" I said above is because you don't need the compiler or unstable_book changes or the feature gate UI test.

See https://github.com/rust-lang/rust/pull/79673/files for an example of adding unstable methods.

jnqnfe commented:
Oh ok, thanks, updated.

CI has passed now finally :D

SkiFire13 commented on library/core/src/cmp.rs:
```diff
@@ -325,6 +325,120 @@ pub enum Ordering {
+    #[must_use]
+    #[unstable(feature = "ordering_helpers", issue = "79885")]
+    pub const fn is_ne(self) -> bool {
+        matches!(self, Less | Greater)
```
`!matches!(self, Equal)` results in better assembly https://godbolt.org/z/M8ojh5 and the same applies to `is_le`. `is_ge` already produces good assembly but for consistency you may want to change that too.

jnqnfe replied:
Ok, implemented! :)

sfackler commented:
@bors r+

bors commented:
:pushpin: Commit 169c59ff0fcbd606c6f639c2318f71b51f6c9207 has been approved by `sfackler`

<!-- @bors r=sfackler 169c59ff0fcbd606c6f639c2318f71b51f6c9207 -->
<!-- homu: {"type":"Approved","sha":"169c59ff0fcbd606c6f639c2318f71b51f6c9207","approver":"sfackler"} -->

bors commented:
:hourglass: Testing commit 169c59ff0fcbd606c6f639c2318f71b51f6c9207 with merge 0c9ef564a7688dd4a9047342c7f8395aea909333...
<!-- homu: {"type":"BuildStarted","head_sha":"169c59ff0fcbd606c6f639c2318f71b51f6c9207","merge_sha":"0c9ef564a7688dd4a9047342c7f8395aea909333"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1535297949)
Approved by: sfackler
Pushing 0c9ef564a7688dd4a9047342c7f8395aea909333 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"sfackler","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1535297949"},"merge_sha":"0c9ef564a7688dd4a9047342c7f8395aea909333"} -->
